src/types/prefix_str.rs

```diff
@@ -46,36 +58,62 @@ macro_rules! prefix_str_mut {
         }
 
         impl<'a> $name<'a> {
+            /// Creates a new reference from a byte slice. This also checks that the str bytes are valid UTF-8.
+            /// To skip this safety check, see [`Self::new_unchecked`].
+            ///
+            /// The `data` is used as the storage for the type.
+            pub fn new(data: &'a mut [u8]) -> Result<Self, std::str::Utf8Error> {
+                // Check that the bytes are valid UTF-8
+                let to_return = unsafe { Self::new_unchecked(data) };
+                std::str::from_utf8(to_return.value)?;
+                Ok(to_return)
+            }
+
             /// Creates a new reference from a byte slice.
             ///
-            /// The `data` is used a the storage for the type.
-            pub fn new(data: &'a mut [u8]) -> Self {
+            /// The `data` is used as the storage for the type.
+            ///
+            /// # Safety
+            /// The caller must guarantee that the resulting str bytes are valid UTF-8.
+            /// The resulting bytes would be the `data` after the size of the length bytes and up to
+            /// the value of the length bytes.
+            pub unsafe fn new_unchecked(data: &'a mut [u8]) -> Self {
                 let type_length = std::mem::size_of::<$prefix_type>();
                 let length = (data.len().saturating_sub(type_length) as $prefix_type).to_le_bytes();
                 data[..type_length].copy_from_slice(&length);
                 Self::from_bytes_mut(data)
             }
 
             /// Loads from a mutable byte slice.
-            pub fn from_bytes_mut(bytes: &'a mut [u8]) -> Self {
+            ///
+            /// # Safety
+            /// The caller must guarantee that the resulting str bytes are valid UTF-8.
+            /// The resulting bytes would be the `data` after the size of the length bytes and up to
+            /// the value of the length bytes.
+            pub unsafe fn from_bytes_mut(bytes: &'a mut [u8]) -> Self {
                 let (length, value) = bytes.split_at_mut(std::mem::size_of::<$prefix_type>());
 
-                let length = bytemuck::from_bytes_mut::<$prefix_type>(length);
-                let value = bytemuck::cast_slice_mut(&mut value[..*length as usize]);
+                let length = bytemuck::pod_read_unaligned::<$prefix_type>(length);
```

Comment:
One issue of using `pod_read_unaligned` is that it creates a copy of the value, so the type won't be (fully) zero-copy. But let's address the alignment issue in a separate PR.